frontend/userpatch.lua

```diff
@@ -0,0 +1,119 @@
+--[[--
+Allows applying developer patches while running KOReader.
+
+The contents in `koreader/userpatches/` are applied on calling `userpatch.applyPatches(priority)`.
+--]]--
+
+local isAndroid, android = pcall(require, "android")
+
+local userpatch =
+    {   -- priorities for user patches,
+        early_once = "0",  -- to be started early on startup (once after an update)
+        early = "1",       -- to be started early on startup (always, but after an `early_once`)
+        late = "2",        -- to be started after UIManager is ready (always)
+                           -- 3-7 are reserved for later use
+        before_exit = "8", -- to be started a bit before exit before settings are saved (always)
+        on_exit = "9",     -- to be started right before exit (always)
+
+        applyPatches = function(priority) end, -- to be overwritten, if the device allows it.
+    }
+
+if isAndroid and android.prop.flavor == "fdroid" then
+    return userpatch -- allows to use applyPatches as a no-op on F-Droid flavor
+end
+
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local DataStorage = require("datastorage")
+
+-- the directory KOReader is installed in (and runs from)
+local package_dir = lfs.currentdir()
+-- the directory where KOReader stores user data (on SDL we may set `XDG_DOCUMENTS_DIR=~/koreader/`)
+local data_dir = os.getenv("XDG_DOCUMENTS_DIR") or DataStorage:getDataDir() or package_dir
```

Comment:
Dunno what XDG_DOCUMENTS_DIR (but it feels a bit generic to be used as `XDG_DOCUMENTS_DIR=~/koreader/` to put in only koreader stuff.
It also makes location a bit inconsistent : most stuff are found in :getDataDir() (ie. user styletweaks), so why allow userpatches to be elsewhere ?
(It would have been more serious if you pass data_dir to shell scripts, as they may not find there what they expect to be in GetDataDir().)

Reply:
I have hijacked XDG_DOCUMENTS_DIR, as this is only used with SDL. So I can have my patch in said dir, not depending on the current git branch. 
But we can introduce a dedicated ENV-variable just to do that (which is definitively not used by anything else). Suggestion: KOREADER_PATCHES_DIR.

Reply:
> So I can have my patch in said dir, not depending on the current git branch.

Don't you have this problem with other stuff lying in data/ like dictionaries, your history.lua ... ?
(Dunno if you're talking about the emulator or AppImage or else - but with my emulator, I expect I would put my patches in:
/koreader/koreader-emulator-x86_64-linux-gnu-debug/koreader/userpatches/
when /koreader/ is the git root.

Reply:
Yepp, it's working to put it in the emulator folder. I just have to remember to update that after a `./kodev clean`.